### DISCUSSION TITLE
Rebasing incredibly slow

### DISCUSSION THREAD

**Discussion by TheDutchCoder** [renovatebot/renovate#11866] [Request Help]
I've raised this issue before, but I'm not sure if I'm doing something wrong.

Whatever I try, I simply can't get renovate to communicate it's rebasing a PR.
Checking the checkbox does nothing. I've waited 15 minutes and it just doesn't do anything. Then after about ~25 minutes it does the rebase, an incredibly long time.

**Comment by TheDutchCoder** [comment]
Okay, it took 33 minutes for it to trigger, that seems excessive.
Is there maybe a way to show on the PR that a rebase is "scheduled", right now there's no feedback whatsoever and people coming from dependabot (like myself), are probably used to direct feedback.

**Reply by rarkins** [reply]
How are you running Renovate, and on what platform?

Edit: looks like github.com hosted app, but will leave this here just to confirm

**Reply by TheDutchCoder** [reply]
Yes it's the github.com hosted app.

**Comment by rarkins** [comment]
Runs should be triggered near instantly when the checkbox is ticked

**Reply by TheDutchCoder** [reply]
Nothing changes when I check the checkbox. It took 33 minutes for it to actually rebase.

There's another PR I currently try to rebase (I checked the box 2 minutes ago now, and no feedback whatsoever, I'll keep you posted on when it actually triggers).

**Reply by viceice** [reply]
https://app.renovatebot.com/dashboard

check the last runs

![image](https://user-images.githubusercontent.com/1798109/134539054-225ae5db-91e4-47eb-b361-6b3537234b23.png)


How long is renovate running on your repo, maybe renovate needs a long time

**Reply by TheDutchCoder** [reply]
![image](https://user-images.githubusercontent.com/1668207/134539253-0ee7b237-d6ce-4dc8-aee7-59d8f57ae842.png)
This is for the previous PR that took 31 minutes (33 was since opening the PR, my bad)

**Reply by rarkins** [reply]
That's pretty close to the timeout

**Reply by TheDutchCoder** [reply]
Yeah the second PR just finished: 26 minutes.
I mean I don't know what the real cause of this is, but let's say it's not very fast ?:)

**Reply by rarkins** [reply]
It's not a PR that finished - it's the repo run. The reason you're not seeing fast rebases is because there's already a job in progress when you check the box, so the second job is queued. 

If you look through one completed run you can maybe work out where the time is going

**Reply by TheDutchCoder** [reply]
All I can see in the logs is that around 45 docker instances are spun up and ran at ~30s to ~60s each, I'm assuming that's due to each package.json file that matches the renovate config. We have quite a lot of lambda's (AWS) that are all their own separate packages in our monorepo so they all get a docker instance.

Those are the only timings in the logs, so I'm assuming two things:
1. This is all done in series, not in parallel, probably compounding the time it takes to go through the repo
2. ~30s to ~60s per docker instance is "normal"

Not sure what we can do about this, maybe multiple renovate configs (is that even a thing)?

I'm sure we're not the first mono-repo users with a medium amount of packages, so maybe there's a proven strategy for this?

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#branchconcurrentlimit

How many branches are created by renovate?
Can you limit it.

What docker images are started by renovate?

**Reply by TheDutchCoder** [reply]
It seems it's doing 20 concurrent branches in the latest run.
I think it's launching docker instances to test if `npm install` works to verify dependency updates are installing fine?
```
DEBUG: Executing command(branch="renovate/proofflow")
{
  "command": [
    "docker run --rm --name=renovate_node --label=renovate_child -v \"/mnt/renovate/gh/creativelayer/creativelayer\":\"/mnt/renovate/gh/creativelayer/creativelayer\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/mnt/renovate/gh/creativelayer/creativelayer/apps/proofflow\" docker.io/renovate/node bash -l -c \"npm i -g npm && hash -d npm && npm install --package-lock-only --no-audit --ignore-scripts\""
  ]
}
```

**Reply by rarkins** [reply]
Best if you reduce the concurrent branches if you have a lot of commits to the main repo and either (a) have enabled `automerge`, or (b) have a "branches must be up to date with the target" branch protection. Either of these things will cause your branches to be rebased every run whenever there's been a commit to the main branch.

**Reply by viceice** [reply]
Renovate will use that to update the npm lockfile. It seems you have a lot deps. So the lockfile update needs too much time.

@rarkins any ideas to dig deeper into it?

**Reply by rarkins** [reply]
If it's rebasing 20 branches per run then that's ~1 minute or so per lock file - not too bad if it's a large monorepo

**Reply by TheDutchCoder** [reply]
We don't commit to `develop` a whole lot on a "per hour" basis, I'd say we merge into `develop` a couple of times a day max.
I can remove automerge, because that doesn't work with reviewers/code-owners anyway, but not sure if that would have any impact.

Again, all I can see is quite a lot of docker spin ups, all taking quite a significant amount of time, resulting in a very, very long process.

I don't know how renovate differs from something like dependabot, I can only give you my end-user experience, but dependabot seemed to always do thing instantaneously, but that might also be because it didn't group PR's for example.

**Reply by rarkins** [reply]
It's somewhat of an apples and oranges comparison. Dependabot supports neither [grouping](https://togithub.com/dependabot/dependabot-core/issues/1190), [automerging](https://togithub.com/dependabot/dependabot-core/issues/1973), or [automatic rebasing](https://togithub.com/dependabot/dependabot-core/issues/3421) of out of date branches. It is these features most likely causing the high run time. 

The reality is that if a tool like Renovate requires you to keep manually requesting rebases, it's not very efficient for you. The problem is you have too many concurrent branches which appear to be rebasing themselves each run. Solutions:

- Decrease the concurrent branch limit (note: this won't go and delete any existing, so won't have an effect until you either merge or close existing ones manually)
- Remove automerge and/or automatic rebasing (set rebaseWhen=conflicted). However if you have branch protection saying PRs must be up to date then it's not ideal to remove automatic rebasing

**Reply by TheDutchCoder** [reply]
Ah okay I see, you guys rebase on every merge to the branch, whereas dependabot only does that in cases of conflicts (or when asked for manually).

I like your approach better, so I will play with the concurrent branch limit a bit.

Just as an aside: even when nothing is currently running and I ask for a rebase with the checkbox, it takes 6 minutes or so before I see anything happening. Is there a way for renovate to (for example how dependabot does it) indicate a PR is queued? Currently you just have to trust things are being processed, whereas dependabot alters the PR body to show that it's currently rebasing and that it might take a while.

Just a better UX basically.

**Comment by TheDutchCoder** [comment]
Right now I'm getting the following issue in the dashboard:

<details>
<summary>INFO log</summary>
<pre>
INFO: Repository started
{
  "renovateVersion": "27.18.0"
}
INFO: Dependency extraction complete
{
  "baseBranch": "develop",
  "stats": {
    "managers": {
      "npm": {
        "fileCount": 69,
        "depCount": 454
      }
    },
    "total": {
      "fileCount": 69,
      "depCount": 454
    }
  }
}
INFO: Branch updated(branch="renovate/amplify")
{
  "commitSha": "8d8af029"
}
INFO: Branch updated(branch="renovate/chalk-a11y")
{
  "commitSha": "d20520d4"
}
INFO: PR updated(branch="renovate/chalk-a11y")
{
  "pr": 2067,
  "prTitle": "chore(deps): update chalk-a11y to v3.2.16"
}
INFO: Branch updated(branch="renovate/chalk-icons")
{
  "commitSha": "c621ed04"
}
INFO: PR updated(branch="renovate/chalk-icons")
{
  "pr": 2071,
  "prTitle": "chore(deps): update dependency vue to v3.2.16"
}
INFO: Branch updated(branch="renovate/graph-api")
{
  "commitSha": "3ac5de17"
}
INFO: Branch updated(branch="renovate/lambda-core")
{
  "commitSha": "67e0d91b"
}
INFO: PR created(branch="renovate/proofflow")
{
  "pr": 2073,
  "prTitle": "chore(deps): update proofflow to v3.2.16"
}
INFO: Branch updated(branch="renovate/sam")
{
  "commitSha": "64140eac"
}
ERROR: Error generating branch cache
{
  "err": {
    "hostType": "github",
    "err": {
      "name": "HTTPError",
      "timings": {
        "start": 1632411863119,
        "socket": 1632411863119,
        "lookup": 1632411863119,
        "connect": 1632411863128,
        "secureConnect": 1632411863137,
        "upload": 1632411863137,
        "response": 1632411863238,
        "end": 1632411863240,
        "phases": {
          "wait": 0,
          "dns": 0,
          "tcp": 9,
          "tls": 9,
          "request": 0,
          "firstByte": 101,
          "download": 2,
          "total": 121
        }
      },
      "message": "Response code 401 (Unauthorized)",
      "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
      "options": {
        "headers": {
          "user-agent": "Renovate Bot (GitHub App 2740)",
          "accept": "application/vnd.github.machine-man-preview+json",
          "authorization": "***********",
          "accept-encoding": "gzip, deflate, br"
        },
        "url": "https://api.github.com/repos/creativelayer/creativelayer/pulls/2073",
        "hostType": "github",
        "username": "",
        "password": "",
        "method": "GET",
        "http2": false
      },
      "response": {
        "statusCode": 401,
        "statusMessage": "Unauthorized",
        "body": {
          "message": "Bad credentials",
          "documentation_url": "https://docs.github.com/rest"
        },
        "headers": {
          "server": "GitHub.com",
          "date": "Thu, 23 Sep 2021 15:44:23 GMT",
          "content-type": "application/json; charset=utf-8",
          "content-length": "80",
          "x-github-media-type": "github.v3; param=machine-man-preview; format=json",
          "x-ratelimit-limit": "60",
          "x-ratelimit-remaining": "55",
          "x-ratelimit-reset": "1632412237",
          "x-ratelimit-used": "5",
          "x-ratelimit-resource": "core",
          "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset",
          "access-control-allow-origin": "*",
          "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
          "x-frame-options": "deny",
          "x-content-type-options": "nosniff",
          "x-xss-protection": "0",
          "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
          "content-security-policy": "default-src 'none'",
          "vary": "Accept-Encoding, Accept, X-Requested-With",
          "x-github-request-id": "9844:7A92:5B033C:635C08:614CA0D7",
          "connection": "close"
        },
        "httpVersion": "1.1"
      }
    },
    "message": "external-host-error",
    "stack": "Error: external-host-error\n    at handleGotError (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/github.js:69:19)\n    at GithubHttp.request (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/github.js:161:13)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async GithubHttp.requestJson (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/index.js:176:21)\n    at async getPr (/home/ubuntu/renovateapp/node_modules/renovate/dist/platform/github/index.js:517:17)\n    at async Proxy.getBranchPr (/home/ubuntu/renovateapp/node_modules/renovate/dist/platform/github/index.js:619:40)\n    at async generateBranchCache (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/cache.js:32:30)\n    at async setBranchCache (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/cache.js:73:26)\n    at async Object.renovateRepository (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/index.js:71:9)\n    at async renovateRepository (/home/ubuntu/renovateapp/app/worker/index.js:400:26)\n    at async /home/ubuntu/renovateapp/app/worker/index.js:644:5"
  },
  "branchName": "renovate/proofflow"
}
WARN: Could not ensure issue
{
  "err": {
    "hostType": "github",
    "err": {
      "name": "HTTPError",
      "timings": {
        "start": 1632411863331,
        "socket": 1632411863331,
        "lookup": 1632411863340,
        "connect": 1632411863348,
        "secureConnect": 1632411863359,
        "upload": 1632411863359,
        "response": 1632411863467,
        "end": 1632411863467,
        "phases": {
          "wait": 0,
          "dns": 9,
          "tcp": 8,
          "tls": 11,
          "request": 0,
          "firstByte": 108,
          "download": 0,
          "total": 136
        }
      },
      "message": "Response code 401 (Unauthorized)",
      "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
      "options": {
        "headers": {
          "user-agent": "Renovate Bot (GitHub App 2740)",
          "accept": "application/vnd.github.machine-man-preview+json",
          "authorization": "***********",
          "content-type": "application/json",
          "content-length": "2807",
          "accept-encoding": "gzip, deflate, br"
        },
        "url": "https://api.github.com/repos/creativelayer/creativelayer/issues",
        "hostType": "github",
        "username": "",
        "password": "",
        "method": "POST",
        "http2": false
      },
      "response": {
        "statusCode": 401,
        "statusMessage": "Unauthorized",
        "body": {
          "message": "Bad credentials",
          "documentation_url": "https://docs.github.com/rest"
        },
        "headers": {
          "server": "GitHub.com",
          "date": "Thu, 23 Sep 2021 15:44:23 GMT",
          "content-type": "application/json; charset=utf-8",
          "content-length": "80",
          "x-github-media-type": "github.v3; param=machine-man-preview; format=json",
          "x-ratelimit-limit": "60",
          "x-ratelimit-remaining": "54",
          "x-ratelimit-reset": "1632412237",
          "x-ratelimit-used": "6",
          "x-ratelimit-resource": "core",
          "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset",
          "access-control-allow-origin": "*",
          "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
          "x-frame-options": "deny",
          "x-content-type-options": "nosniff",
          "x-xss-protection": "0",
          "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
          "content-security-policy": "default-src 'none'",
          "vary": "Accept-Encoding, Accept, X-Requested-With",
          "x-github-request-id": "9846:837A:678898:745741:614CA0D7",
          "connection": "close"
        },
        "httpVersion": "1.1"
      }
    },
    "message": "external-host-error",
    "stack": "Error: external-host-error\n    at handleGotError (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/github.js:69:19)\n    at GithubHttp.request (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/github.js:161:13)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async GithubHttp.requestJson (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/index.js:176:21)\n    at async Proxy.ensureIssue (/home/ubuntu/renovateapp/node_modules/renovate/dist/platform/github/index.js:921:9)\n    at async ensureDependencyDashboard (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/dependency-dashboard.js:282:9)\n    at async Object.renovateRepository (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/index.js:81:13)\n    at async renovateRepository (/home/ubuntu/renovateapp/app/worker/index.js:400:26)\n    at async /home/ubuntu/renovateapp/app/worker/index.js:644:5"
  }
}
INFO: Repository finished
{
  "durationMs": 1799770
}
</pre>
</details>

**Reply by rarkins** [reply]
I don't know the cause, but that log indicates that a valid token is not being used. The auth limit is 60.

**Comment by TheDutchCoder** [comment]
Let me also post our config, in case that might be useful:

<details>
<summary>Renovate Config</summary>
<pre>
{
  "extends": [
    "config:base"
  ],
  "hostRules": [
    {
      "matchHost": "https://npm.pkg.github.com/",
      "hostType": "npm",
      "encrypted": {
        "token": "TOKEN_HERE"
      }
    }
  ],
  "npmrc": "@creativelayer:registry=https://npm.pkg.github.com",
  "includePaths": [ "amplify/**", "sam/**", "apps/**", "packages/**", "workflow" ],
  "labels": ["dependencies"],
  "packageRules": [
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "amplify/**/package.json" ],
      "groupName": "amplify",
      "automerge": true
    },
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "sam/**/package.json" ],
      "groupName": "sam",
      "automerge": true
    },
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "apps/proofflow/package.json" ],
      "groupName": "proofflow",
      "addLabels": ["proofflow"],
      "ignoreDeps": ["jest", "babel-jest", "babel-preset-jest", "ts-jest"]
    },
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "workflow/package.json" ],
      "groupName": "workflow",
      "addLabels": ["workflow"],
      "ignoreDeps": ["jest", "babel-jest", "babel-preset-jest", "ts-jest"]
    },
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "packages/chalk-icons/package.json" ],
      "groupName": "chalk-icons",
      "automerge": true,
      "addLabels": ["chalk-icons"]
    },
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "packages/chalk-a11y/package.json" ],
      "groupName": "chalk-a11y",
      "automerge": true,
      "ignoreDeps": ["jest", "babel-jest", "babel-preset-jest", "ts-jest"]
    },
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "packages/chalk-ui/package.json" ],
      "groupName": "chalk-ui",
      "automerge": true,
      "addLabels": ["chalk-ui"],
      "ignoreDeps": ["jest", "babel-jest", "babel-preset-jest", "ts-jest"]
    },
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "packages/graph-api/package.json" ],
      "groupName": "graph-api",
      "automerge": true
    },
    {
      "matchPackagePatterns": [ "*" ],
      "matchPaths": [ "packages/lambda-core/package.json" ],
      "groupName": "lambda-core",
      "automerge": true
    }
  ]
}
</pre>
</details>
